tools: Add scripts for makinga new release

pevik opened the pull request:
Adding two scripts: tools/tag-release.sh and tools/create-tarballs.sh.
See: https://github.com/iputils/iputils/issues/479

pevik commented:
@nmeyerhans @vapier @robert-scheck Can you please have a look?

vapier commented on tools/lib.sh:
```diff
@@ -0,0 +1,31 @@
+
+rod()
+{
+	eval "$@" || quit "$@ failed"
```
yikes, don't use `eval`.  it's not even necessary here -- just delete it.

also, use `$*`, not `$@` when calling `quit`

pevik replied:
> yikes, don't use eval. it's not even necessary here -- just delete it.

This is inspired by LTP, which uses `tst_rod` binary. I'm not sure if everything would work without `eval`, but maybe you're right.

vapier commented on tools/lib.sh:
```diff
@@ -0,0 +1,31 @@
+	exit 1
+}
+
+rod()
```
i have no idea what this is short for.  func (all of them really) could use a short comment header.

pevik replied:
+1. Inspired by LTP, I wonder myself.

nmeyerhans replied:
The best I could come up with when reading this was that it's short for "run-or-die".  But yeah, some more descriptive names and/or descriptive comments would be nice.

pevik replied:
Yep, it is "run or die". I added descriptive comments, although it's a bit overkill for 1 or 2 lines script :).

vapier replied:
it's easy to write bad shell.  but bad shell is also bad ;).

pevik commented on tools/create-tarballs.sh:
```diff
@@ -0,0 +1,54 @@
+dist=builddir
+formats="xztar,gztar,zip"
+
+. $(dirname "$0")/lib.sh
```
I usually check just with checkbashisms, which catches a real problems.
Some of the `shellcheck` errors does not make sense:
`SC3043 (warning): In POSIX sh, 'local' is undefined.`

or does not make sense in the context:
`^--^ SC2162 (info): read without -r will mangle backslashes.` (when `read answer` compares only against `[Yy]`

This is IMHO invalid (will not work on dash or busybox sh - these might be used on alpine):
`SC2010 (warning): Don't use ls | grep. Use a glob or a for loop with a condition to allow non-alphanumeric filenames.`

Also not sure if `prevent word splitting` is feasible to fix - with `rod` it needs to be escaped.
But sure, some warnings and errors are valid, thus fixing at least some of them (more will likely follow, just not sure if I fix all of them).

vapier replied:
while `local` is not defined in POSIX, every shell in Linux supports it, so it's not something i would bother with

pevik commented on tools/create-tarballs.sh:
```diff
@@ -0,0 +1,54 @@
+rm -f *.sha256sum
+
+for file in *.*; do
+	sha256sum $file >> sha256sums
```
No, 1) list of the files is evaluated before `sha256sums` is created 2) `*.*` pattern would not select it anyway.

vapier replied:
sorry, i thought it was `"*"`, not `*.*`

at this point, why is there a loop at all ?  just run the command directly once.  it produces the same output.

nmeyerhans reviewed (commented):
I had some comments queued up, but @vapier beat me to them.  +1 to everything he said.  Especially around:
1. Ensure variables referencing path names are quoted
2. Get rid of `eval`
3. `shellcheck` looks like it reports a few legitimate concerns; It would be good to fix them.

Looks pretty good in general, though.  I like the release notes autogeneration.

pevik commented:
> I had some comments queued up, but @vapier beat me to them. +1 to everything he said. Especially around:
> 
> 1. Ensure variables referencing path names are quoted
I've personally haven't heart of anybody who would clone git repository to the path which contains a space, which is IMHO the only reason why quoting is needed (or am I missing something?), but sure, I'll fix it.

> 2. Get rid of `eval`
Done.

> 3. `shellcheck` looks like it reports a few legitimate concerns; It would be good to fix them.
Sure.

> Looks pretty good in general, though. I like the release notes autogeneration.

Thanks for your feedback.

pevik commented:
@vapier @nmeyerhans thanks for your review. I updated the docs + slightly changed the scripts.

I dared to keep these shellcheck info/warnings:
```
In create-tarballs.sh line 44:
rm -f *.sha256sum
      ^-- SC2035 (info): Use ./*glob* or -- *glob* so names with dashes won't become options.

In create-tarballs.sh line 55:
mv -v * "$dir"
      ^-- SC2035 (info): Use ./*glob* or -- *glob* so names with dashes won't become options.

In create-tarballs.sh line 59:
cd "$dir" && ls -p | grep -v /
             ^-- SC2010 (warning): Don't use ls | grep. Use a glob or a for loop with a condition to allow non-alphanumeric filenames.

In tag-release.sh line 51:
git shortlog -s -n "$old_tag".. >> "$credit"
^-- SC2129 (style): Consider using { cmd1; cmd2; } >> file instead of individual redirects.


In lib.sh line 8:
	local msg="$1"
        ^-------^ SC3043 (warning): In POSIX sh, 'local' is undefined.


In lib.sh line 9:
	local answer
        ^----------^ SC3043 (warning): In POSIX sh, 'local' is undefined.
```

vapier commented on tools/tag-release.sh:
```diff
@@ -0,0 +1,79 @@
+fi
+
+if ! git --no-pager diff --exit-code; then
+	quit "Please commit your changes before making new release"
```
this doesn't handle changes staged in the index, or unknown files

pevik replied:
Would you add something else? Maybe
```
if ! git --no-pager diff --exit-code || ! git --no-pager diff --staged  --exit-code; then
```

Fell free to post a command. BTW I didn't meant this check to be perfect, it's just a sanity check.

Maybe not being on master branch could be a problem worth to detect (we probably don't want to release from some random branch, release is a single commit on already widely tested master branch, but merging from other branch could merge more commits to master or fail during push to master):

```
if [ "$(git branch --show-current)" != "master" ]; then
    quit "not on master branch"
fi
```
```

vapier replied:
personally i don't have release scripts operate in the current checkout at all.  git makes it cheap to create a new shared checkout using the current repo.  so check that out to a temp location, do the generation/commit/tag steps in there, and then throw it away.

pevik replied:
That's already in `tools/create-tarballs.sh`, because that's really safer for creating tarballs (to have just source files). But I thought the tag creation should be ok work in the current checkout. Sure, I can clone with `-b master`, but then I might need to ask user to verify 1) if the latest commit on master is the expected one to tag, 2) preferably then pull the release commit and tag in the working directory (not strictly needed for the release).

pevik commented:
I fixed 2 more warnings (`Use ./*glob* or -- *glob* so names with dashes won't become options.`) and added above mentioned git checks.

I also added checks for non-basic commands being installed (git, meson, ...), so that script fails at the beginning, not in the middle of the work.

@vapier @nmeyerhans I dared to add your `Reviewed-by` tags for first and second commit, because you spent time with reviewing. Feel free to have final check on them + please check for the doc change (the grammar, as I'm not a native speaker). I suppose we spent really a lot of time on this, let's get it merged.
